libsrc/target/aquarius/stdio/generic_console_ioctl.asm

```diff
@@ -1,126 +1,148 @@
-    ; VCTRL_MODE_TILEMAP   = (1 << 1),
-    ; VCTRL_MODE_BITMAP    = (2 << 1),
-    ; VCTRL_MODE_MASK      = (3 << 1),
-    ; VCTRL_SPRITES_ENABLE = (1 << 3),
-    ; VCTRL_TEXT_PRIORITY  = (1 << 4),
-    ; 1 = mode 0  = text 40x25
-    ; 2 = mode 1 - 64x32 tilemap
-    ; 3 = mode 2  = hires 320x200
-    ; +8 = sprites enable
-    ; +16 = text priority
-    ;
-    ; If text is enabled, we preferentially print to there
-    ; 
-
-    ld      (__aquarius_mode),a
-    out     (PORT_VCTRL),a
-    ld      c,CLIB_GENCON_CAPS_TEXT
-    ld      b,24
-    bit     1,a
-    jr      nz,set_caps
-    ld      c,CLIB_GENCON_CAPS_BITMAP
-    ld      b,25
+        cp      IOCTL_GENCON_SET_MODE
+        jr      nz, failure
+        ld      a, e
+        ; VCTRL_MODE_OFF             = (0 << 1),
+        ; VCTRL_TEXT_ENABLE          = (1 << 0),
+        ; VCTRL_MODE_TILEMAP         = (1 << 1),
+        ; VCTRL_MODE_BITMAP (1bpp)   = (2 << 1),
+        ; VCTRL_MODE_MC (4bpp)       = (3 << 1),
+        ; VCTRL_SPRITES_ENABLE       = (1 << 3),
+        ; VCTRL_TEXT_PRIORITY        = (1 << 4),
+        ; 1 = mode 0  = text 40x25
+        ; 2 = mode 1  = 64x32 tilemap
+        ; 3 = mode 2  = hires 320x200 1bpp
+        ; 4 = mode 3  = lores 160x200 4bpp
+        ; +8   = sprites enable
+        ; +16  = text priority
+        ; +32  = remap border color
+        ; +64  = 80 column text mode
+        ; +128 = select 2nd text page
+        ;
+        ; If text is enabled, we preferentially print to there
+        ;
+
+        ld      (__aquarius_mode), a
+        out     (PORT_VCTRL), a
+
+        ld      b, a
+        bit     6, a
+        ld      a, 40
+        jr      z, setWidth
+        ld      a, 80
+setWidth:
+        ld      (__console_w), a
+        ld      a, b
+
+        ld      c, CLIB_GENCON_CAPS_TEXT
+        bit     0, a
+        jr      nz, set_caps
+        ld      c, CLIB_GENCON_CAPS_BITMAP
 set_caps:
-    ld      a,b
-    ld      (__console_h),a
-    ld      a,c
-    ld      (generic_console_caps),a
-    call    generic_console_cls
-    and	    a
-    ret
+        ld      a, 25
+        ld      (__console_h), a
+        ld      a, c
+        ld      (generic_console_caps), a
+        call    generic_console_cls
+        and     a
+        ret
 failure:
-    scf
-    ret
+        scf
+        ret
+
+
+        SECTION code_crt_init
 
+        EXTERN  set_default_palette
 
-    SECTION code_crt_init
+        ; On an Aquarius+ we modify the caps so that we can define the font
+        ld      c, CLIB_AQUARIUS_PLUS
+        rr      c
+        jr      nc, not_plus
 
-    EXTERN  set_default_palette
+        call    set_default_palette
 
-    ; On an Aquarius+ we modify the caps so that we can define the font
-    ld      c,CLIB_AQUARIUS_PLUS
-    rr      c
-    jr      nc,not_plus
-    ld      a,CLIB_GENCON_CAPS_TEXT
-    ld      (generic_console_caps),a
-    call    set_default_palette
+        ; Remap the border color character for aqplus
```

Comment:
@suborb . It's not showing here for some reason, but I added the code to code_crt0_init section so the hardware and the library should get sync'd during init.

```
        SECTION code_crt_init

        EXTERN  set_default_palette

        ; On an Aquarius+ we modify the caps so that we can define the font
        ld      c, CLIB_AQUARIUS_PLUS
        rr      c
        jr      nc, not_plus

        call    set_default_palette

        ; Remap the border color character for aqplus
        in      a, (IO_VCTRL)
        or      VCTRL_REMAP_BC
        ld      e, a
        ld      a, IOCTL_GENCON_SET_MODE
        ; This will setup the mode and screen size
        ; so that it matches the hardware during init.
        call    ck_mode
```

Reply:
The `code_crt_init` section is only included if any of the `PUBLIC` definitions in the file are required.

What this means is that this initialisation is only included if the user calls `console_ioctl()`  (probably an existing issue looking at the palette issue)

Reply:
Ohhh, got it. I was thinking it would be pulled in if --generic-console was used on the command line.

Any thoughts on how to handle this? I'll also take a look at some of the other targets to see how they handle this.

Reply:
I’ve tended to put them into generic_console.asm to handle that case. 

You’ll need to move the CAPS defc out  as